document offline cht server (#449)
medic/cht-docs: content/en/apps/guides/hosting/offline.md

---

> title: "Offline CHT Server"
> linkTitle: "Offline Hosting"
> weight: 100
> description: >
>  Running an instance of CHT Core server with no Internet  connection

Reviewer: My comments didn't seem to make it to the review -- posting again:

```suggestion
title: "Offline Hosting of CHT Server"
linkTitle: "Offline Hosting"
weight: 100
description: >
 Deploying and hosting CHT Core server instances without Internet connectivity
```

---

{{% alert title="Note" %}} This guide is not meant for a production CHT instance.  Support may be added in the future an offline CHT server in a production environment.  Please see the "Considerations" section below. {{% /alert %}}

## Introduction

The CHT is built as an offline first application. This applies to clients, either  browsers or Android applications, connecting to the CHT server.  The server itself assumes it has Internet connectivity to provide services such as DNS, software updates and general use connectivity.  This document explores what it looks like when the CHT server is offline without these services available.

Running a CHT server offline requires no modifications to the CHT itself.  Instead, supporting services normally found online are replicated locally. 

## Considerations

An offline CHT server is most appropriate for a development environment.  There are serious implications to consider before deploying an offline instance per our [existing requirements]({{< relref "requirements#considerations" >}}).

> Additionally, if users are going to migrate between offline locations with the same domain name, always ensure a different login and password is used. This will prevent a client from another CHT instance trying to synchronize with a CHT instance it shouldn't synchronize with, possibly causing data corruption.

Author: nice catch on the privacy aspect!

## Requirements

 A CHT instance is accessible offline when you can resolve the domain to an IP address, and a TLS certificate is on the CHT server with a common name (CN) that matches the domain name. On top of the [existing requirements]({{< relref "requirements" >}}), also needed are:

### Static IP

The CHT server needs to be given a static IP so that DNS will always resolve to the correct host.

### TLS Certificate

Browsers might allow you to connect to a server with an invalid TLS certificate after you [bypass the warning](https://www.ssl.com/guide/troubleshooting-ssl-tls-browser-errors-and-warnings/).  This is not so for an Android app like [Medic Android App](https://github.com/medic/medic-android/) - it requires a valid TLS certificate to work correctly.  You will need to acquire a valid TLS certificate from a certificate authority (CA) and install it on your CHT server.

> It is common to use [Let's Encrypt](https://en.wikipedia.org/wiki/Let%27s_encrypt) for this because they provide free certificates. Be cautioned that Let's Encrypt certificates expire after 90 days, so the server will need to be constantly updated with a new certificate.  Other CAs provide certificates that expire after a year, so this concern will always apply.

Author: Cool cool - the "this" clarity and lighter language make sense. thanks!

After acquiring the certificate, if you are running a docker based CHT deployment, see [TLS instructions for Docker]({{< relref "/apps/guides/hosting/ssl-cert-install" >}}) to install the certificate. 

### Domain Name Server

In order to match the static IP of the web server to the CN in the certificate, a Domain Name Server (DNS) must be used.  This will allow any client on the LAN to easily connect to your CHT server without needing anything more than the domain name. 

Most LANs will defer to the Internet Service Provider (ISP) to provide DNS, but there is no ISP in an offline scenario.  Instead, one must be provided. This DNS server will then be configured to have an `A` record (or `AAAA` in the case of IPv6) to point to the CHT server.

### Dynamic Host Configuration Protocol 

Any new client that connects to a network will get an IP address from a Dynamic Host Configuration Protocol (DHCP) server.  It is critical that the DHCP server for the LAN the CHT is on instructs all clients to use the DNS server configured above.  

> This is the final piece for any client, specifically Android clients, on the LAN to successfully connect to the CHT server and correctly use its TLS certificate.  

Author: yup yup - I think it was the final piece in an early draft, but this paragraph doesn't add much anyway!

### Wi-Fi AP 

> Optionally, a Wi-FI Access Point (AP) may be deployed on the LAN.  This should not be a router, but a simple AP so as  not introduce a DHCP or DNS server that could conflict with the two configured above. An AP will enable a device without Ethernet, such as an Android phone, to easily connect to the CHT server.

Author: From myopically technical perspective it's not needed, but you're right - the CHT's purpose in life is to be an offline, in the field, CHW used app - not an in the oflice, always online in a full browser SaaS!

APs don't need to be routers (and vice versa). In striving to simplify this sentence, I can see how I've complicated things a bit mire!

I'll re-word this and see if I can improve things.

> Optionally, a Wi-FI Access Point (AP) may be deployed on the LAN.  This should not be a router, but a simple AP so as  not introduce a DHCP or DNS server that could conflict with the two configured above. An AP will enable a device without Ethernet, such as an Android phone, to easily connect to the CHT server.

Author: fixed in 23321c2

### Expertise

> Deployments that don't have staff familiar with DNS, TLS Certs, DHCP, LAN topology and Linux in general should likely not use an offline CHT server. This is a complex deployment where mistakes are easy to make unless proper training is in place. 

Author: fixed in 23321c2

## Benefits Over Other Solutions

An offline deployment may consider substituting some requirements above with these other solutions.  Note that ngrok and local-ip.co require Internet connectivity, so are not an offline solution.

### ngrok

> Aside from the obvious benefit of not requiring Internet connectivity, when an offline solution is deployed, traffic stays 100% local.  When using either [your own reverse proxy]({{< relref "/secure-sharing-of-developer-instance" >}}) or a third party provider like [ngrok](https://ngrok.com/), traffic may traverse 100s or 1,000s of kilometers to ultimately reach the CHT server which is 10 meters away.
> 
> This can help even in locations with Internet connectivity which is very slow, very expensive per megabyte, or both.

Reviewer: Worth simplifying a bit, and avoiding the orphaned "this".

```suggestion
When an offline solution is deployed, traffic stays 100% local, whereas when using either [your own reverse proxy]({{< relref "/secure-sharing-of-developer-instance" >}}) or a third party provider like [ngrok](https://ngrok.com/), traffic may traverse 100s or 1,000s of kilometers to ultimately reach the CHT server which is 10 meters away. This can help when Internet connectivity is very slow, very expensive per megabyte, or both.
```
